Scroll cursor into view after pull

finkef opened the pull request:
Fixes #2022.

This PR fixes scroll issues when navigating to thoughts that need to be pulled from storage.

Since depending on the amount of thoughts that need to be pulled, [a single animation frame previously used](https://github.com/cybersemics/em/blob/b20fd295186c0b323c036c70bfa635e9da4a58c5/src/device/scrollCursorIntoView.ts#L63-L67) to wait until the loaded thoughts have been rendered is not enough and has been replaced with a timeout.
Before any programmatic scroll occurs, we disable reacting to scroll events to not falsely identify a programmatic scroll as a user scroll. Unfortunately, there is no way of reliably detecting whether a scroll event is triggered by a user or by JavaScript.

raineorshine commented on src/device/scrollCursorIntoView.ts:
```diff
@@ -37,13 +38,15 @@ const scrollIntoViewIfNeeded = (el: Element | null | undefined) => {
   const scrollYNew = isAboveViewport
-    ? y - (toolbarRect?.height ?? 0) - rect.height / 2
-    : y - viewport.innerHeight + viewport.virtualKeyboardHeight + rect.height * 1.5
+    ? y - (toolbarRect?.height ?? 0) - rect.height / 2 + 40
+    : y - viewport.innerHeight + viewport.virtualKeyboardHeight + rect.height * 1.5 + 40
```
Can you open a separate PR for the NavBar offset? It's not essential for this issue, and I'd like to track the solution separately since it is a fix for #1827.

FYI The height of the NavBar is not always 40px. We'll want to measure the height like we measure the toolbar height.

raineorshine commented on src/device/scrollCursorIntoView.ts:
```diff
@@ -37,13 +38,15 @@ const scrollIntoViewIfNeeded = (el: Element | null | undefined) => {
   const y = window.scrollY + rect.y
 
   // leave a margin between the element and the viewport edge equal to half the element's height
+  // add 40px to the bottom margin to prevent the element from being hidden by the footer
```
Thanks for this! Can you open a separate PR (at your hourly rate)? It's not essential for this issue, and I'd like to track the solution separately since it is a fix for #1827.

FYI The height of the NavBar is not always 40px. We'll want to measure the height like we measure the toolbar height.

finkef replied:
Pushed a new PR here: https://github.com/cybersemics/em/pull/2301

raineorshine commented on src/redux-middleware/pullQueue.ts:
```diff
@@ -178,6 +180,11 @@ const pullQueueMiddleware: ThunkMiddleware<State> = ({ getState, dispatch }) =>
     syncStatusStore.update({ isPulling: false })
     pulling.delete(extendedPullQueueFiltered)
 
+    // scroll cursor into view if the pull was the result of navigation to a thought
+    if (navigatedStore.getState()) {
+      scrollCursorIntoView()
+    }
+
```
It would be better to move this into the [scrollCursorIntoView middleware](https://github.com/cybersemics/em/blob/bbb87809aba6cbb002cdc3cdfa4c300f45cf2b9c/src/redux-middleware/scrollCursorIntoView.ts), so the logic is more centralized, and the pullQueue doesn't need to know about the navigatedStore or cursor.

We can't directly detect when a `pull` completes from the other middleware, but we have a couple other options:

1. Since a `pull` always dispatches [updateThoughts](https://github.com/cybersemics/em/blob/bbb87809aba6cbb002cdc3cdfa4c300f45cf2b9c/src/actions/updateThoughts.ts), you can call `scrollCursorIntoView` in response by checking if `action.type === 'updateThoughts'` . This was suggested in the original issue. Did you run into problems with this approach?
2. Since a `pull` is always followed by `syncStatusStore.update({ isPulling: false })`, you can subscribe to the `syncStatusStore` and call `scrollCursorIntoView` when `isPulling` changes to false.

raineorshine commented on src/device/scrollCursorIntoView.ts:
```diff
@@ -57,14 +60,15 @@ const scrollCursorIntoView = () => {
+  setTimeout(() => {
     scrollIntoViewIfNeeded(document.querySelector('.editing'))
-  })
+  }, 100)
```
Can we add the 100ms only when needed? 100ms is going to cause a noticeable delay in a lot of other cases.

raineorshine commented on src/stores/navigated.ts:
```diff
@@ -0,0 +1,31 @@
+ */
+export const preventScrollDetection = () => {
+  isAutoScrolling = true
+  setTimeout(() => {
+    isAutoScrolling = false
+  }, 200)
```
This timeout should probably be persisted and cancelled at the beginning of the function. Otherwise two calls within 200ms will cause `isAutoScrolling` to flip back-and-forth.

The timeout should probably also be cancelled when the user navigates.

raineorshine commented on src/stores/navigated.ts:
```diff
@@ -0,0 +1,31 @@
+  isAutoScrolling = true
+  setTimeout(() => {
+    isAutoScrolling = false
+  }, 200)
```
Where does the 200ms come from? Is that meant to match the [duration of the browser's smooth scroll](https://stackoverflow.com/a/71417475/480608)?

finkef replied:
No, this was the lowest number of ms that were needed to prevent interrupting the scroll, but I believe this was a consequence from a previous attempt (based on updateThoughts, which is triggered quite often). Although giving it a single frame is enough from my testing right now, I've left a delay of 50ms in case there are still some conflicting calls.
The only way that this would affect the UX is if a user navigates to a thought and starts scrolling within 50ms, in this case, the programmatic scroll would still take precedence.

raineorshine reviewed (changes requested):
I've tested this out, and it seems to work quite well, so I'm happy with the overall approach.

It seems I overengineered the solution with my suggestion of a new ministore. A ministore is only really useful if a real-time subscription is required. Otherwise, a simple [global](https://github.com/cybersemics/em/blob/bbb87809aba6cbb002cdc3cdfa4c300f45cf2b9c/src/globals.ts) variable will do. Unless I'm missing something, we're just reading + writing this state in a couple places, but never subscribing.

finkef commented:
Refactored and contained everything within `scrollCursorIntoView` and its middleware. Much better, thanks for the feedback!

raineorshine commented on src/redux-middleware/scrollCursorIntoView.ts:
```diff
@@ -19,6 +24,26 @@ editingValueStore.subscribe(
+ * Scroll the cursor into view if the sync was the result of navigation to a thought.
+ */
+syncStatusStore.subscribe(state => {
+  if (!state.isPulling && navigated) {
```
We only want this triggered when `isPulling` *changes* to false, not every time it is false. The `syncStatusStore` can update for other reasons.

You can either modify the ministore to pass the old state to `subscribe`, or you can just add an `isPullingPrev` variable here to hold onto the previous value.

raineorshine commented on src/device/scrollCursorIntoView.ts:
```diff
@@ -44,11 +69,19 @@ const scrollIntoViewIfNeeded = (el: Element | null | undefined) => {
   // scroll to 1 instead of 0
   // otherwise Mobile Safari scrolls to the top after MultiGesture
   // See: touchmove in MultiGesture.tsx
+  preventScrollDetection()
```
Is `preventScrollDetection` intended to be called with every invocation of `window.scrollTo`, or only the ones in `scrollCursorIntoView`?

finkef replied:
Only the ones in `scrollCursorIntoView`. The idea here is to not cancel itself when called multiple times as thoughts are loaded.

raineorshine reviewed (changes requested):
Great, this is looking better.

Do we need both `autoScrolling` and `navigated`? I'm concerned that there are three different states for this one feature: `autoScrolling`, `navigated`, and `isNavigated`. If there is any way to simplify that would help the complexity.

Analogously, we have a profusion of concepts here: "autoscrolling", "programmatic scrolling", and "preventing scroll detection". Do these really all refer to different things? All of them are fine names, but I think there's some overlap there. We should simplify if possible. For example, we could rename `preventScrollDetection` to `startAutoscroll` so that we eliminate the concept of "scroll detection".

finkef commented:
> Do we need both `autoScrolling` and `navigated`? I'm concerned that there are three different states for this one feature: `autoScrolling`, `navigated`, and `isNavigated`. If there is any way to simplify that would help the complexity.
> 
> Analogously, we have a profusion of concepts here: "autoscrolling", "programmatic scrolling", and "preventing scroll detection". Do these really all refer to different things? All of them are fine names, but I think there's some overlap there. We should simplify if possible. For example, we could rename `preventScrollDetection` to `startAutoscroll` so that we eliminate the concept of "scroll detection".

Yes, we do need both. `navigated` indicates that the cursor changed and we should scroll as a result. `autoScrolling` is to make sure multiple invocations don't cancel themselves as we receive scroll events that are indistinguishable between a user and a programmatic scroll.
`isNavigated` is only introduced to not introduce a dependency cycle, but for consistency's sake, I'll change it to `navigated`. That better indicates that we're talking about the same concept.

finkef commented:
Adapted naming.

raineorshine commented on src/redux-middleware/scrollCursorIntoView.ts:
```diff
@@ -19,6 +24,30 @@ editingValueStore.subscribe(
+ * Scroll the cursor into view if the sync was the result of navigation to a thought.
+ */
+syncStatusStore.subscribe(state => {
+  if (!state.isPulling && prevIsPulling && navigated) {
```
Any reason not to call `scrollCursorIntoView` whenever `isPulling` changes to false, regardless of `navigated`? It would reduce the complexity.

finkef replied:
That would basically tie pulling to scrolling, not sure if there are cases in which we don't want to force a scroll, e.g. when loading favorites?

raineorshine commented on src/redux-middleware/scrollCursorIntoView.ts:
```diff
@@ -27,7 +56,10 @@ const scrollCursorIntoViewMiddleware: ThunkMiddleware<State> = ({ getState }) =>
+      // indicate that the cursor has changed and we want to scroll it into view
+      // this is needed for when thoughts need to be pulled from storage prior to scrolling
+      navigated = true
+      scrollCursorIntoView({ navigated: true })
```
Setting `navigated: true` here is too general. In most cases, we want to scroll the cursor into view on the next animation frame, not after 100ms (really, as soon as the hierarchy re-renders).

Under what conditions do we need to delay 100ms? Did changing `requestAnimationFrame to `setTimeout(..., 0)` break anything?

finkef replied:
`requestAnimationFrame` and `setTimeout(..., 0)` worked the same, otherwise would have opted to stick with rAF in an if/else.

raineorshine reviewed (changes requested):
I don't think we're quite there yet. I suspect that there is a way to simplify or streamline this solution. Right now it has too many moving parts.

First, we can move `device/scrollCursorIntoView.ts` completely into `redux-middleware/scrollCursorIntoView`, as it is not used anywhere else. .This was the case before this PR, but it became more obvious in this implementation. The export `isAutoscrolling` can be removed completely once all the logic shares the same scope.

Second, I just realized that the term "autoscrolling" already has a meaning within the app. It refers to when the browser automatically scrolls to a focused input element. See: https://github.com/cybersemics/em/blob/9ce879fbff5481c6d30a125774d6c21009b23406/src/device/preventAutoscroll.ts. So we should rename `autoScrolling` and related variables to something like `scrollingCursorIntoView`, `scrollingToCursor`, or `forcedScroll`.

Once those changes are made, see if you see any other opportunities to simplify the logic or reduce state.

finkef commented:
I've refactored the code, put everything into the middleware and changed the naming of variables for more clarity.
Most notably, I've inverted the behavior of `navigated` and called it `userInteractedAfterNavigation`, I believe this better reflects what it's actually doing.

I've changed the added `setTimeout` to wait for `LAYOUT_NODE_ANIMATION_DURATION` (instead of 100ms, this was wrong actually) after a pull. The timeout is necessary to get the correct scroll position of the off-screen element since its computed position is influenced by (other) nodes animating to their final positions.
Because of this, fully omitting the timeout is not possible. I've tried instead of relying on the computed rect to use the actual top style value from its parent `[aria-label="tree-node"]` to workaround requiring a timeout, but this does not work due to `scrollTo` being limited by the actual viewport height which is not extended at this point.

Fully removing `navigated/userInteractedAfterNavigation` is not possible/advisable in my opinion since we can't otherwise have a user scrolling take precedence over these programmatic scrolls.

raineorshine commented on src/redux-middleware/scrollCursorIntoView.ts:
```diff
@@ -19,6 +114,30 @@ editingValueStore.subscribe(
+ */
+scrollTopStore.subscribe(() => {
+  // Ignore scroll events that are triggered programmatically.
+  if (forcedScrolling) return
```
One way to eliminate `forcedScrolling` would be to unsubscribe from `scrollTopStore` on `startForcedScrolling`, and then resubscribe after the 50ms. That would reduce the amount of state, but would add complexity to the subscription.

Seems to be about the same level of complexity all things considered.

raineorshine reviewed (approved):
I'm happy with this. There are still a lot of moving parts, but having it all contained in one middleware helps a lot. The use of `LAYOUT_NODE_ANIMATION_DURATION` definitely seems correct.

Thanks for your persistence!
